arch/Mips/MipsMapping.c

```diff
@@ -1,1071 +1,454 @@
+	Mips_inc_op_count(MI);
+}
 
-const char *Mips_group_name(csh handle, unsigned int id)
+static void Mips_set_detail_op_operand(MCInst *MI, unsigned OpNum)
 {
-#ifndef CAPSTONE_DIET
-	return id2name(group_name_maps, ARR_SIZE(group_name_maps), id);
-#else
-	return NULL;
-#endif
+	cs_op_type op_type = map_get_op_type(MI, OpNum) & ~CS_OP_MEM;
+	if (op_type == CS_OP_IMM) {
+		Mips_set_detail_op_imm(MI, OpNum,
+					    MCInst_getOpVal(MI, OpNum));
+	} else if (op_type == CS_OP_REG) {
+		Mips_set_detail_op_reg(MI, OpNum,
+					    MCInst_getOpVal(MI, OpNum));
+	} else
+		printf("Operand type %d not handled!\n", op_type);
+}
+
+static void Mips_set_detail_op_branch(MCInst *MI, unsigned OpNum)
+{
+	cs_op_type op_type = map_get_op_type(MI, OpNum) & ~CS_OP_MEM;
+	if (op_type == CS_OP_IMM) {
+		uint64_t Target = (uint64_t)MCInst_getOpVal(MI, OpNum);
+		Mips_set_detail_op_imm(MI, OpNum, Target + MI->address);
+	} else if (op_type == CS_OP_REG) {
+		Mips_set_detail_op_reg(MI, OpNum,
+					    MCInst_getOpVal(MI, OpNum));
+	} else
+		printf("Operand type %d not handled!\n", op_type);
+}
+
+static void Mips_set_detail_op_uimm(MCInst *MI, unsigned OpNum)
+{
+	Mips_set_detail_op_imm(MI, OpNum,
+				    MCInst_getOpVal(MI, OpNum));
+}
+
+static void Mips_set_detail_op_uimm_offset(MCInst *MI, unsigned OpNum,
+											unsigned Bits, uint64_t Offset)
+{
+	uint64_t Imm = MCInst_getOpVal(MI, OpNum);
+	Imm -= Offset;
+	Imm &= (1 << Bits) - 1;
+	Imm += Offset;
+	Mips_set_detail_op_imm(MI, OpNum, Imm);
+}
+
+static void Mips_set_detail_op_mem_nanomips(MCInst *MI, unsigned OpNum)
+{
+	if (!detail_is_set(MI) || !doing_mem(MI))
+		return;
+
+	MCOperand *Op = MCInst_getOperand(MI, OpNum);
+	Mips_get_detail_op(MI, 0)->type = MIPS_OP_MEM;
+	// Base is a register, but nanoMips uses the Imm value as register.
+	Mips_get_detail_op(MI, 0)->mem.base = MCOperand_getImm(Op);
+	Mips_get_detail_op(MI, 0)->access = map_get_op_access(MI, OpNum);
 }
 
-// map instruction name to public instruction ID
-mips_reg Mips_map_insn(const char *name)
+static void Mips_set_detail_op_reglist(MCInst *MI, unsigned OpNum, bool isNanoMips)
 {
-	// handle special alias first
-	unsigned int i;
+	if (!detail_is_set(MI))
+		return;
 
-	// NOTE: skip first NULL name in insn_name_maps
-	i = name2id(&insn_name_maps[1], ARR_SIZE(insn_name_maps) - 1, name);
+	if (isNanoMips) {
+		for (unsigned i = OpNum; i < MCInst_getNumOperands(MI); i++) {
+			Mips_set_detail_op_reg(MI, i, MCInst_getOpVal(MI, i));
+		}
+		return;
+	}
+	// -2 because register List is always first operand of instruction
+	// and it is always followed by memory operand (base + offset).
+	for (unsigned i = OpNum, e = MCInst_getNumOperands(MI) - 2; i != e; ++i) {
+		Mips_set_detail_op_reg(MI, i, MCInst_getOpVal(MI, i));
+	}
+}
 
-	return (i != -1)? i : MIPS_REG_INVALID;
+static void Mips_set_detail_op_uimm_address(MCInst *MI, unsigned OpNum)
+{
+	uint64_t Target = MI->address + (uint64_t)MCInst_getOpVal(MI, OpNum);
```

Comment:
Same as above. Check if address should be added at all.

Reply:
this is a copy of what the printing instructions do. otherwise i need to use args to get this value (`address`).

Reply:
See above. Didn't thought it to an end properly.